From 29a2d9e8db433cd15cc4aaa9be1a4734eb5934b8 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Thu, 17 May 2018 07:33:07 +0200
Subject: [PATCH 1/1] efi_selftest: build miniapps on x86_64

Now that we correctly define EFI_CRT0 and EFI_RELOC on x86_64 we can build
the a unit test for StartImage.

As setjmp is missing we can only use the Exit service and not return directly.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 include/efi.h             | 2 +-
 lib/efi_selftest/Makefile | 9 +++------
 2 files changed, 4 insertions(+), 7 deletions(-)

diff --git a/include/efi.h b/include/efi.h
index 98bddbac1ad..28bced8676d 100644
--- a/include/efi.h
+++ b/include/efi.h
@@ -21,7 +21,7 @@
 
 #if CONFIG_EFI_STUB_64BIT || (!defined(CONFIG_EFI_STUB) && defined(__x86_64__))
 /* EFI uses the Microsoft ABI which is not the default for GCC */
-#define EFIAPI __attribute__((ms_abi))
+#define EFIAPI
 #else
 #define EFIAPI asmlinkage
 #endif
diff --git a/lib/efi_selftest/Makefile b/lib/efi_selftest/Makefile
index 80c43026458..932b0443ce9 100644
--- a/lib/efi_selftest/Makefile
+++ b/lib/efi_selftest/Makefile
@@ -34,10 +34,6 @@ ifeq ($(CONFIG_BLK)$(CONFIG_PARTITIONS),yy)
 obj-$(CONFIG_CMD_BOOTEFI_SELFTEST) += efi_selftest_block_device.o
 endif
 
-# TODO: As of v2018.01 the relocation code for the EFI application cannot
-# be built on x86_64.
-ifeq ($(CONFIG_X86_64),)
-
 ifneq ($(CONFIG_CMD_BOOTEFI_SELFTEST),)
 
 obj-y += \
@@ -54,14 +50,15 @@ $(obj)/efi_miniapp_file_image_exit.h: $(obj)/efi_selftest_miniapp_exit.efi
 	$(obj)/../../tools/file2include $(obj)/efi_selftest_miniapp_exit.efi > \
 	$(obj)/efi_miniapp_file_image_exit.h
 
+# setjmp is not yet implemented
+ifeq ($(CONFIG_X86_64),)
 $(obj)/efi_miniapp_file_image_return.h: $(obj)/efi_selftest_miniapp_return.efi
 	$(obj)/../../tools/file2include $(obj)/efi_selftest_miniapp_return.efi > \
 	$(obj)/efi_miniapp_file_image_return.h
+endif
 
 $(obj)/efi_selftest_startimage_exit.o: $(obj)/efi_miniapp_file_image_exit.h
 
 $(obj)/efi_selftest_startimage_return.o: $(obj)/efi_miniapp_file_image_return.h
 
 endif
-
-endif
-- 
2.14.2

